apps/Test App/Test App/View Models/UploadCreationViewModel.swift

```diff
@@ -0,0 +1,226 @@
+//
+//  UploadCreationViewModel.swift
+//  Test App
+//
+//  Created by Emily Dixon on 5/14/23.
+//
+
+import Foundation
+import PhotosUI
+import MuxUploadSDK
+
+class UploadCreationViewModel : ObservableObject {
+    
+    struct PickerError: Error {
+        
+        static var unexpectedFormat: PickerError {
+            PickerError(localizedDescription: "Unexpected video file format")
+        }
+        
+        static var missingAssetIdentifier: PickerError {
+            PickerError(localizedDescription: "Missing asset identifier")
+        }
+        
+        static var createUploadFailed: PickerError {
+            PickerError(localizedDescription: "Upload could not be created")
+        }
+        
+        var localizedDescription: String
+        
+    }
+    
+    func requestPhotosAccess() {
+        switch photosAuthStatus {
+        case .cant_auth(_): logger.critical("This application can't ask for permission to access photos. Check your Info.plist for NSPhotoLibraryAddUsageDescription, and make sure to use a physical device with this app")
+        case .authorized(_): logger.warning("requestPhotosAccess called but we already had access. ignoring")
+        case .can_auth(_): doRequestPhotosPermission()
+        }
+    }
+    
+    func startUpload(preparedMedia: PreparedUpload, forceRestart: Bool) -> MuxUpload {
+        let upload = MuxUpload(
+            uploadURL: preparedMedia.remoteURL,
+            videoFileURL: preparedMedia.localVideoFile
+        )
+        upload.progressHandler = { progress in
+            self.logger.info("Uploading \(progress.progress?.completedUnitCount ?? 0)/\(progress.progress?.totalUnitCount ?? 0)")
+        }
+        upload.resultHandler = { result in }
+        upload.start(forceRestart: true)
+        return upload
+    }
+    
+    /// Prepares a Photos Asset for upload by exporting it to a local temp file
+    func tryToPrepare(from pickerResult: PHPickerResult) {
+        // Cancel anything that was already happening
+        if let assetRequestId = assetRequestId {
+            PHImageManager.default().cancelImageRequest(assetRequestId)
+        }
+        if let prepareTask = prepareTask {
+            prepareTask.cancel()
+        }
+        if let thumbnailGenerator = thumbnailGenerator {
+            thumbnailGenerator.cancelAllCGImageGeneration()
+        }
+        
+        // TODO: This is a very common workflow. Should the SDK be able to do this workflow with Photos?
```

Comment:
Yes I think it should, I'll be committing a change that handles this workflow in the SDK as part of the input standardization changes

Reply:
🙌 🙌